CRM/Core/Permission/Joomla.php

```diff
@@ -131,4 +135,74 @@ public function checkGroupRole($array) {
     return FALSE;
   }
 
+  /**
+   * @inheritDoc
+   */
+  public function upgradePermissions($permissions) {
```

Comment:
I'm adding some inline comments, but I'm not going to change the name of the array. I'm pretty sure doing so would throw strict warnings about the child method changing the signature of the parent method. And if I were to change the parent method, then I'd have to do the same for all of the children.